compose: Have at least 2 new lines before and after a quoted message.

N-Shar-ma opened the pull request:
Uptil now, 1 new line was added before and 1 after a quoted message. Now for more breathing room around a quoted message, new lines are inserted to space it from any content before and after by at least 2 new lines.

Fixes: #23608.


**Screenshots and screen captures:**

https://user-images.githubusercontent.com/68962290/214903141-7c043c2e-bf75-4e65-b2d8-7640af94b375.mp4



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
I think there was some discussion on the previous PR about whether it'd make sense to make this a parameter passed to `insert_syntax_and_focus`, rather than something the quote-and-reply logic did -- did you consider that possibility?

alya commented:
Works for me in manual testing!

N-Shar-ma commented:
> I think there was some discussion on the previous PR about whether it'd make sense to make this controlled by a parameter passed to `insert_syntax_and_focus`, rather than something the quote-and-reply logic did -- did you consider that possibility?
> 
> One motivation is we might want to use that parameter to resolve #11389 too.

Hmm I did have a reason when I pushed the commit 1st, but in hindsight, after more thought, I realise it was rather weak (basically I hadn't realised I could - and should! - refactor `quote_and_reply` beyond passing in a "block" parameter, which limited the utiliity of that approach in my view then). Now I have changed the code to make it parameter controlled, and it works the same as before (manually tested) but the node tests are yet to be fixed

N-Shar-ma commented:
The more I work on this code, the more unexpected subtleties I notice... The tests too seem like they will be messy to fix... not sure if this is high priority enough for the time investment 🤔

alya commented:
Let's pause on this if it's tricky. I'll remove the "integration review" label.

alya commented:
It would be fine for another contributor to take over the PR and fix the tests.

shenderaj9021 commented:
May I work on this issue?

N-Shar-ma commented on web/tests/compose_ui.test.js:
```diff
@@ -279,23 +279,25 @@ run_test("quote_and_reply", ({override, override_rewire}) => {
-            console.info(arg);
-            throw new Error("We expected the actual code to pass in a string.");
-        }
+        // Not used right now, but could be in future.
```
Since the `quote_and_reply` function now only uses `caret` with either no argument or a number passed in, all this commented out code never runs and has lost test coverage. I wasn't sure if I should remove this, hence have commented it for now

N-Shar-ma commented:
@alya I have fixed the tests and added new ones for the new behaviour. The only changes are in the test files, so I don't think this needs manual testing again. @timabbott Could you please review this again?

timabbott commented:
Finally found the time to read this carefully; this looks great. Can you try the suggested approach for that commented code, and then we can merge this!

timabbott commented:
I added a bit more of comment by the `istanbul ignore` line, and marked this to merge once CI passes.

I seem to recall there were some potential follow-ups to make other compose actions use this new function as well; can you perhaps bump the chat.zulip.org thread for this change with an analysis of which callers exist and which you think might want to pass the new option?

Thanks @N-Shar-ma!
